wmfsm/wmfsm.1
.\"                                      Hey, EMACS: -*- nroff -*-
.TH WMFSM 1 "Jan 13, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
wmfsm \- shows the amount of disk space available
.SH SYNOPSIS
.BR wmfsm " [ " -bhfn " | " --help " | [ " --normal " | " --fire " ] [ " --blink " | " --noblink " ] [ " -display
.IR display " ]"
.RB " [ " --delay " | " -d
.IR delay " ] ]"
.SH DESCRIPTION
\fBWmfsm\fP is a WindowMaker dock application that shows the amount of disc
space available on mounted filesystems with coloured bars.
.SH OPTIONS
.TP
.B \-\-normal, \-n
Show the entire bars in the same colour.
.TP
.B \-\-fire, \-f
Render the bars as gradients. You may switch between the normal and fire modes
by mouse clicks anywhere on \fBwmfsm\fP.
.TP
.BR \-\-blink, \-b " | " \-\-noblink
Do / do not blink if a filesystem is 95% full.
.TP
.BI \-display \ display
Draw to
.BR X (1)
display \fIdisplay\fP.
.TP
.B \-d, \-\-delay \ \fIdelay\fP
Wait \fIdelay\fP milliseconds to redraw the window.
.TP
.B \-h, \-\-help
Show summary of options.
.SH FILES
You may exclude filesystems from being shown by \fBwmfsm\fP by listing them
in the ~/.wmfsmrc file. This file has to look like this:
.IP
.nf
[exclude]
/dev/pts
.fi
.SH SEE ALSO
.BR X (1x),
.BR wmaker (1),
.BR df (1).
.SH AUTHOR
This manual page was written by Arthur Korn <arthur@korn.ch> and
modified by Chris Gray <cgray@tribsoft.com>


debian/changelog
wmfsm (0.35-1) unstable; urgency=medium

  * New upstream release.
  * New maintainer (Closes: #776014).
  * debian/compat
    - Bump to 9.
  * debian/control
    - Bump Standards-Version to 3.9.6.
    - Bump versioned dependency on debhelper to >=9.
    - Add Homepage and Vcs-*.
    - Remove unnecessary Build-Depend x11proto-core-dev.
    - Replace autotools-dev with dh-autoreconf in Build-Depends.
  * debian/copyright
    - Add DEP5 header, debian/*, and wmgeneral/* sections.
  * debian/docs
    - Add AUTHORS.
  * debian/patches
    - (handle_HOME_errors.patch) Avoid errors related to HOME environment
      variable.
  * debian/rules
    - Update to use dh with dh_autoreconf.
    - Add get-orig-source target.
  * debian/source/format
    - New file; use 3.0 (quilt).
  * debian/watch
    - New file.

 -- Doug Torrance <dtorrance@monmouthcollege.edu>  Thu, 22 Jan 2015 14:47:21 -0600

wmfsm (0.34-13) unstable; urgency=low

  * QA upload.
  * Replace deprecated build-dependency on x-dev with x11proto-core-dev
    (Closes: #515405).
  * Add ${misc:Depends} substvar to binary package.
  * Use "Applications/System/Administration" menu section instead of
    "Apps/System".
  * debian/rules
    - Don't ignore errors in clean target.
    - Don't use deprecated "dh_clean -k".
  * Update debian/copyright.
  * Bump Standards-Version to 3.8.2.
  * Move to debhelper 7.

 -- Chris Lamb <lamby@debian.org>  Tue, 28 Jul 2009 15:18:19 +0200

wmfsm (0.34-12) unstable; urgency=low

  * Orphan: set maintainer to QA.

 -- Arthur Korn <arthur@debian.org>  Mon, 24 Nov 2008 19:39:17 +0100

wmfsm (0.34-11) unstable; urgency=low

  * Update Standards-Version to 3.7.2.2 without changes.

 -- Arthur Korn <arthur@debian.org>  Mon, 12 Feb 2007 21:22:47 +0100

wmfsm (0.34-10) unstable; urgency=low

  * configure: Rebuild with autoconf 2.59a-7 to fix tests for X11
    libraries (Closes: #347639, #346889).

 -- Arthur Korn <arthur@debian.org>  Sat, 14 Jan 2006 15:40:46 +0100

wmfsm (0.34-9) unstable; urgency=low

  * control: Replace build-dependency on xlibs-dev by detailed
    build-dependencies according to the xlibs-split tool
    (Closes: #346889).

 -- Arthur Korn <arthur@debian.org>  Mon,  9 Jan 2006 19:31:15 +0100

wmfsm (0.34-8) unstable; urgency=low

  * rules: Fix timestamps of various autoconfig related files to
    avoid unnecessary rebuild (and the dependency on autoconf)
    (Closes: #321011).

 -- Arthur Korn <arthur@debian.org>  Wed,  3 Aug 2005 01:02:58 +0200

wmfsm (0.34-7) unstable; urgency=low

  * Remove build dependency on automake1.6
  * Add build dependency on autotools-dev and no update config.{sub,guess} in
    clean target, correctly call ./configure.
  * Bump Standards-Version to 3.6.2.0 without changes.
  * Update to debhelper compat level 4 with minor changes.
  * docs: Remove NEWS (that file is empty?!) and TODO (out of date).

 -- Arthur Korn <arthur@debian.org>  Sun, 31 Jul 2005 23:48:07 +0200

wmfsm (0.34-6) unstable; urgency=low

  * Fix filesystem name corruption. Thanks to Ralf Horstmann and Cristian
    Ionescu-Idbohrn for the patch.

 -- Arthur Korn <arthur@korn.ch>  Wed, 25 Aug 2004 01:29:58 +0200

wmfsm (0.34-5) unstable; urgency=low

  * wmfsm.sample: Adapt to the wealth of pseudo-fs in current linux
    systems (Closes: #226298).
  * Remove debian/watch, upstream's dead.
  * wmfsm.1: Make a point on the strictness of wmfsmrc format (Closes: #226274).
  * Memleak news: avoid it if there are more than 9 filesystems too.
  * Bumpt to Standards-Version 3.6.1.0 without changes.

 -- Arthur Korn <arthur@korn.ch>  Wed, 17 Mar 2004 13:04:33 +0100

wmfsm (0.34-4) unstable; urgency=low

  * Fix memleak introduced in 0.34-2 with the patch by Timo Benk.
    global char *mp[] was assigned the result of stdup() repeatedly
    without free. My ugly fix is to free it all up at the beginning of
    readFileSystems(). Now I'm a proud Perl coder ... (Closes: #196057).

 -- Arthur Korn <arthur@korn.ch>  Wed,  2 Jul 2003 23:05:21 +0200

wmfsm (0.34-3) unstable; urgency=low

  * Run aclocal and update config.sub and config.guess
    (Closes: #174686, #187602).

 -- Arthur Korn <arthur@korn.ch>  Wed,  2 Jul 2003 23:04:55 +0200

wmfsm (0.34-2) unstable; urgency=low

  * Add Build-Depends on automake1.6 (Closes: #175082).
  * Update config.{sub,guess} from autotools-dev.
  * Apply patch by Timo Benk that makes wmfsm reread the filesystems list
    during its life (Closes: #186826).

 -- Arthur Korn <arthur@korn.ch>  Tue,  1 Apr 2003 11:25:11 +0200

wmfsm (0.34-1) unstable; urgency=low

  * New upstream release.
  * rules: New DEB_BUILD_OPTIONS syntax.
  * Bump Standards-Version to 3.5.8.
  * Build upstream changelog by concatenating Changelog{,.old}.

 -- Arthur Korn <arthur@korn.ch>  Tue, 17 Dec 2002 11:10:00 +0100

wmfsm (0.33-2) unstable; urgency=low

  * Correct spelling error in description (Closes: #125491).

 -- Arthur Korn <arthur@korn.ch>  Tue, 18 Dec 2001 22:13:56 +0100

wmfsm (0.33-1) unstable; urgency=low

  * New upstream release.
    - autoconf

 -- Arthur Korn <arthur@korn.ch>  Sat, 20 Jan 2001 16:06:04 +0100

wmfsm (0.31-9) unstable; urgency=low

  * Add menu hints.

 -- Arthur Korn <arthur@korn.ch>  Sun, 14 Jan 2001 17:06:04 +0100

wmfsm (0.31-8) unstable; urgency=low

  * Adapt Build-Depends to new libs (see changelog of last version).

 -- Arthur Korn <arthur@korn.ch>  Sat, 18 Nov 2000 21:42:29 +0100

wmfsm (0.31-7) unstable; urgency=low

  * Recompile without libc5 dependency (libxpm4.7) (Closes: #77277).

 -- Arthur Korn <arthur@korn.ch>  Fri, 17 Nov 2000 22:25:44 +0100

wmfsm (0.31-6) unstable; urgency=low

  * control: Change maintainer address to arthur@debian.org.
  * This time including full source in upload.

 -- Arthur Korn <arthur@korn.ch>  Wed, 15 Nov 2000 20:36:47 +0100

wmfsm (0.31-5) unstable; urgency=low

  * wmfsm.1: Write a manpage.
  * control: Change section to utils.

 -- Arthur Korn <arthur@korn.ch>  Sun,  8 Oct 2000 10:43:21 +0200


wmfsm/wmfsm.c
/*
 *
 *  	wmfsm-0.33 (C) 1999 Stefan Eilemann (Stefan.Eilemann@dlr.de)
 * 
 *  		- Shows file system usage ala mfsm
 * 
 * 
 * 
 *
 * 	This program is free software; you can redistribute it and/or modify
 * 	it under the terms of the GNU General Public License as published by
 * 	the Free Software Foundation; either version 2, or (at your option)
 * 	any later version.
 *
 * 	This program is distributed in the hope that it will be useful,
 * 	but WITHOUT ANY WARRANTY; without even the implied warranty of
 * 	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * 	GNU General Public License for more details.
 *
 * 	You should have received a copy of the GNU General Public License
 * 	along with this program (see the file COPYING); if not, write to the
 * 	Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
 *      Boston, MA  02111-1307, USA
 *
 *
 *
 *      Changes: #include "../ChangeLog" 
 *
 */





/*  
 *   Includes  
 */
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <X11/X.h>
#include <X11/xpm.h>
#include "../wmgeneral/wmgeneral.h"
#include "wmfsm_master.xpm"
#include "wmfsm_mask.xbm"

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifdef SVR4
#include <netdb.h>
#include <sys/systeminfo.h>
#endif /* SVR4 */

#ifdef HAVE_SYS_VFS_H
#include <sys/vfs.h>
#endif
#ifdef HAVE_SYS_STATFS_H
#include <sys/statfs.h>
#endif
#ifdef HAVE_SYS_MOUNT_H
#include <sys/mount.h>
#endif
#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif

#ifdef STATFS_2_ARGUMENTS
# define STATFS(a,b) statfs(a,b)
#elif defined STATFS_4_ARGUMENTS
# define STATFS(a,b) statfs(a,b,sizeof(struct statfs),0)
#else
# define STATFS(a,b) statfs(a,b) /* Maybe configure got messed up */
#endif


/*
#if defined IRIX64
# include <sys/types.h>
# define STATFS(a,b) statfs(a,b,sizeof(struct statfs),0)
#elif defined linux 
# define STATFS(a,b) statfs(a,b)
#elif  defined(SunOS)
# define STATFS(a,b) statfs(a,b,sizeof(struct statfs),0)
#elif defined(__OpenBSD__) || defined(__FreeBSD__)
# define STATFS(a,b) statfs(a,b)
# include <sys/param.h>
#endif		
*/
 
#define _GNU_SOURCE
#include <getopt.h>
/* 
 *  Delay between refreshes (in microseconds), uses the 10*DELAY_10
 *   coz irix has max of 100000L :(. 
 */
#define DELAY_10 99999L
#define WMFSM_VERSION "0.33"
#define LENMP 5	/*max 10, number char for mountpoint*/

/*modes for drawing*/
#define NORMAL 0
#define FIRE 1


void ParseCMDLine(int argc, char *argv[]);
void pressEvent(XButtonEvent *xev);
void readFileSystems(void);
void excludeFileSystems(void);
void print_usage(void);

int		UpToDate = 0;
int		ForceUpdate = 1;
char	*mp[100];
int		per[9];
int		numberfs;
int		mode = FIRE;
int		blink = 1;
int		blinkper = 95;
char dummy[4096];
char    *myName;
int delay = DELAY_10;

int xpos[]={ 66,71,76,81,86,91, /* A B C D E F */
			 66,71,76,81,86,91, /* G H I J K L */
			 66,71,76,81,86,91, /* M N O P Q R */
			 66,71,76,81,86,91, /* S T U V W X */
			 66,71,76,81,86,91, /* Y Z / _ - . */
             96,101,			/* 0 1 */
             96,101,			/* 2 3 */
             96,101,			/* 4 5 */
             96,101,			/* 6 7 */
             96,101 };			/* 8 9 */

int ypos[]={  4, 4, 4, 4, 4, 4,
			  9, 9, 9, 9, 9, 9,
             14,14,14,14,14,14,
             19,19,19,19,19,19,
             24,24,24,24,24,24,
              4, 4,
              9, 9,
			 14,14,
  			 19,19,
			 24,24 };


/*  
 *   main  
 */
int main(int argc, char *argv[]) 
{

    XEvent  event;
    int     dx,dy;
    char    hostname[100];
    int		i,j,k;
    int     c,on[9]={1,1,1,1,1,1,1,1,1};
    struct statfs buffer;

    /*
     *  Parse any command line arguments.
     */
    myName=strdup(argv[0]);
    ParseCMDLine(argc, argv);
    readFileSystems();   
    
	   
    openXwindow(argc, argv, wmfsm_master_xpm, wmfsm_mask_bits, wmfsm_mask_width, wmfsm_mask_height);


#ifndef SVR4
    if (gethostname(hostname, 100) != 0) {
        perror("gethostname");
	exit(10);
    }
#else
    if (sysinfo(SI_HOSTNAME, hostname, 100) < 0) {
	perror("sysinfo(SI_HOSTNAME)");
	exit(20);
    }
#endif
    while(1) {
	    /* 
	     *   Process any pending X events.
	     */
        while(XPending(display)){
            XNextEvent(display, &event);
            switch(event.type){
                case Expose:
                        RedrawWindow();
                        break;
                case ButtonPress:
                        pressEvent(&event.xbutton);
                        mode = (mode+1)%2;
                        break;
                case ButtonRelease:
                        break;
            }
        }
	    /*! Here read data
	     */
        for (i=0;i<numberfs;i++){
            if (STATFS(mp[i],&buffer)!=-1){
                per[i]=100-(int)((float)buffer.f_bfree/(float)buffer.f_blocks*100.0);
            }
        }
	    /*!
	     *  Draw window.
	     */
	    /*
	     * Clear window.
	     */
	    copyXPMArea(5, 69, 54, 54, 5, 5);
	    /*! here draw values
	     * void copyXPMArea(int x, int y, int sx, int sy, int dx, int dy)
	     */
         
		 if (numberfs>4){
	    	for (i=0,dy=0;i<numberfs;i++){
            	for (j=0,dx=0;j<LENMP && j<strlen(mp[i]); j++){
                    k = j+(strlen(mp[i])>LENMP ? strlen(mp[i])-LENMP  : 0);
                    c = (int)mp[i][k];
                	switch(c){
            	    	case '/': 
                	    	c=123;break;
            	    	case '_': 
                	    	c=124;break;
                    	case '0':case '1':case '2':case '3':case '4':
                    	case '5':case '6':case '7':case '8':case '9':
                    		c+=79;break;
            	    	default:
				/* All _tolower()s are now tolower()
				 * -- tschroed */
                	    	c=tolower(c);break;
                	}
                	copyXPMArea(xpos[c-97], ypos[c-97], 5, 5, 6+dx, 5+dy);dx+=5;
		    	}
            	c = (int)((float)per[i]*0.52+66.0);
            	k = (10-LENMP)*5+2;
            	k = (float)per[i]/100.0*k;
            	k = LENMP*5+k;
            	k = k < LENMP*5+1 ? LENMP*5+1 : k;
            	c = c > 66 ? c : 66;
            	c = c > 117 ? 117 : c;
                if( blink ){
                    if ( (int)per[i] >= blinkper )
                        on[i] = !on[i];
                    else 
                        on[i] = 1;
                } else
                    on[i] = 1;
                
            	for(j=LENMP*5;j<k;j++){
                    if (mode==FIRE && on[i])
        	    	    copyXPMArea(66+(float)(j-LENMP*5)*52.0/(float)((10-LENMP)*5),29,1,5,6+j,5+dy);
                    else if (mode==NORMAL && on[i])
        	    	    copyXPMArea(c,29,1,5,6+j,5+dy);
                }
            	dy+=6;
	    	}
		}else{ /*one fs in two lines*/
	    	for (i=0,dy=0;i<numberfs;i++){
            	for (j=0,dx=0;j<10 && j<strlen(mp[i]); j++){
        	    	c = (int)mp[i][j+(strlen(mp[i])>10 ? strlen(mp[i])-10 : 0)];
                	switch(c){
            	    	case '/': 
                	    	c=123;break;
            	    	case '_': 
                	    	c=124;break;
                    	case '0':case '1':case '2':case '3':case '4':
                    	case '5':case '6':case '7':case '8':case '9':
                    		c+=79;break;
            	    	default:
                	    	c=tolower(c);break;
                	}
                	copyXPMArea(xpos[c-97], ypos[c-97], 5, 5, 6+dx, 5+dy);dx+=5;
		    	}
            	dy+=6;
            	c = (int)((float)per[i]*0.52+66.0);
            	k = (float)per[i]/100.0*52;
            	k = k < 1 ? 1 : k;
            	c = c > 66 ? c : 66;
            	c = c > 117 ? 117 : c;
                if( blink ){
                    if ( (int)per[i] >= blinkper )
                        on[i] = !on[i];
                    else 
                        on[i] = 1;
                } else
                    on[i] = 1;

            	for(j=0;j<k;j++){
                	if (mode==FIRE && on[i])
        	    		copyXPMArea(66+(float)(j)*52.0/50.0,29,1,5,6+j,5+dy);
                    else if (mode==NORMAL && on[i])
        	    		copyXPMArea(c,29,1,5,6+j,5+dy);
                }
            	dy+=6;
	    	}
		}
        if (numberfs<9)
            for (j=0,dx=0,dy=47;j<10 && j<strlen(hostname); j++){
                c = (int)hostname[j];
                switch(c){
        	        case '/': 
            	        c=123;break;
        	        case '_': 
            	        c=124;break;
        	        case '-': 
            	        c=125;break;
        	        case '.': 
            	        c=126;break;
                    case '0':case '1':case '2':case '3':case '4':
                    case '5':case '6':case '7':case '8':case '9':
                	    c+=79;break;
        	        default:
            	        c=tolower(c);break;
                }
                copyXPMArea(xpos[c-97], ypos[c-97], 5, 5, 6+dx, 6+dy);dx+=5;
		    }
        	
	    RedrawWindow();
	    /* 
	     *  Wait for next update 
	     */
	    i=0;
	    while(i++<10) usleep(delay);
	}
}



/*
 *   ParseCMDLine()
 */
void ParseCMDLine(int argc, char *argv[]) 
{

	int c, option_index, arg;
	static struct option long_options[] =
	{
		{"fire", no_argument, 0, 'f'},
		{"normal", no_argument, 0, 'n'},
		{"blink", no_argument, &blink, 1},
		{"noblink", no_argument, &blink, 0},
		{"delay", required_argument, 0, 'd'},
                {"help", no_argument, 0, 'h'},
		{0, 0, 0, 0}
	};
	while(1){
		c = getopt_long(argc, argv, "bd:fhn", long_options, &option_index);
		if(c == -1)
			break;
		switch(c){
			case 0: /* If blink or noblink was toggled */
				break;
			case 'b':
				blink = 1;
				break;
			case 'd':
				delay = atoi(optarg);
				break;
			case 'f':
				mode = FIRE;
				break;
                        case 'h':
                                print_usage();
                                exit(0);
			case 'n':
				mode = NORMAL;
				break;
			case '?':
				break;
			default:
				print_usage();
				exit(1);
		}
	}
}


void print_usage()
{

    printf("\nwmfsm version: %s\n", WMFSM_VERSION);
    printf("\nusage: wmfsm \n");
    printf("\t--normal, -n\t\tDraw bars in normal mode.\n");
    printf("\t--fire, -f\t\t\tDraw bars in fire mode.\n");
    printf("\t--[no]blink\t\tBlinks if a filesystem is 95 percent full.\n");
    printf("\t-display <Display>\tUse alternate X display.\n");
    printf("\t--delay <number>, -d\tUse a delay that is not the default.\n");
    printf("\t-h\t\t\tDisplay help screen.\n");
}

void pressEvent(XButtonEvent *xev)
{
   ForceUpdate = 1;
   return;
}

void readFileSystems()
{
/* Look for the goods between #if defined(__OpenBSD__) -- tschroed */
#if defined(__OpenBSD__) || defined(__FreeBSD__)
#define MAXMOUNT	32
    struct statfs sfs[MAXMOUNT];
    int fscount;
    
/* OpenBSD has no /etc/mtab, we use getfsstat instead -- tschroed */
    if((fscount=getfsstat(sfs,sizeof(sfs),0))==-1)
    {
	    perror("getfsstat");
	    exit(1);
    }
    for(numberfs=0;numberfs<fscount;numberfs++)
	/* We won't watch RO mounts. */
	if(sfs[numberfs].f_flags & MNT_RDONLY)
	    { numberfs--; fscount--; }
	else
	    mp[numberfs]=strdup(sfs[numberfs].f_mntonname);
#else /* __OpenBSD__ || __FreeBSD__ */

    FILE    *fp;
    char    mountPoint[255],dummy[255],fstype[255],options[255];
    
#if defined(SunOS)
        /* Solaris uses /etc/mnttab */
    fp=fopen("/etc/mnttab","r");
    if (!fp){
        fprintf(stderr,"%s:Can't open /etc/mnttab for reading\n",myName);
        exit (1);
    }
#else
    fp=fopen("/etc/mtab","r");
    if (!fp){
        fprintf(stderr,"%s:Can't open /etc/mtab for reading\n",myName);
        exit (1);
    }
#endif    
    
    numberfs = 0;
    while (!feof(fp) && numberfs<100){
#if defined(SunOS)
          /* only five entries per row in /etc/mnttab */
        fscanf(fp,"%s %s %s %s %s\n",dummy,mountPoint,fstype,options,dummy);
#else
        fscanf(fp,"%s %s %s %s %s %s\n",dummy,mountPoint,fstype,options,dummy,dummy);
#endif    
        
        if (
#if defined IRIX64 || defined(SunOS)
            strcmp(fstype,"hwgfs") && 
            strcmp(fstype,"autofs") && 
            strcmp(fstype,"proc") && 
            strcmp(fstype,"fd") && 
            !strstr(options,"ignore")
#elif defined linux
            strcmp(fstype,"proc") &&
	    strcmp(fstype, "shm")
#else
            1
#endif		
        ){
            mp[numberfs++]=strdup(mountPoint);
        }
    }
#endif /* __OpenBSD__ || __FreeBSD__ */
    excludeFileSystems();
}
void excludeFileSystems()
{
    char confFileName[255];
    char workString[255];
    int  i,j,exnumberfs=0;
    int  start = 0, excluded, finalnumberfs=0;
    char *exmp[100];
    FILE *confFile;

	strncpy ( confFileName , (char *)getenv("HOME"),245 );
	strcat ( confFileName , "/.wmfsmrc" );
    confFile = fopen(confFileName, "r");
    if (confFile){
        while (!feof(confFile)){
            if (!fgets(workString,255,confFile)) break;
            if (workString[strlen(workString)-1] == '\n')
                workString[strlen(workString)-1] = '\0';
            /* convert all chars readed to lower , for case insensitive strcmp() */
            /* look for start block token [xxx] */
            if ( workString[0] == '[' )
                /* look for starting license block in .simpackrc */
                if ( !strcmp(workString,"[exclude]") )  start = 1;
                else                                    start = 0;
            /* look for feature sring in license block */
            if ( start && strcmp(workString,"[exclude]") ){
                exmp[exnumberfs++]=strdup(workString);
            }
        }
    }else{
        numberfs=numberfs >9 ? 9 : numberfs;
        return;
    }
    fclose(confFile);
    /* in exmp are now all filesystems to exclude, now do it */
    if (!exnumberfs){
        numberfs=numberfs >9 ? 9 : numberfs;
        return;
    }
    for (i=0;i<numberfs;i++){
        for (j=0,excluded=0;j<exnumberfs;j++){
            if (!strcmp(mp[i],exmp[j]))
                excluded=1;
        }
        if (!excluded)
            mp[finalnumberfs++]=strdup(mp[i]);
    }
    for (j=0;j<exnumberfs;j++)
        free(exmp[j]);
    for (j=finalnumberfs;j<numberfs;j++)
        free(mp[j]);
    numberfs= finalnumberfs>9 ? 9 : finalnumberfs;
}
